clients/admin-ui/src/features/auth/auth.slice.ts

```diff
@@ -0,0 +1,101 @@
+import {
+  createListenerMiddleware,
+  createSlice,
+  PayloadAction,
+} from '@reduxjs/toolkit';
+import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';
+
+import type { RootState } from '../../app/store';
+import { STORED_CREDENTIALS_KEY } from '../../constants';
+import { User } from '../user-management/types';
+import { LoginRequest, LoginResponse } from './types';
+
+export interface AuthState {
+  user: User | null;
+  token: string | null;
+}
+
+const initialState: AuthState = {
+  token: null,
+  user: null,
+};
+
+// Auth slice
+export const authSlice = createSlice({
+  name: 'auth',
+  initialState,
+  reducers: {
+    login(
+      state,
+      { payload: { user_data, token_data } }: PayloadAction<LoginResponse>
+    ) {
+      return Object.assign(state, {
+        user: user_data,
+        token: token_data.access_token,
+      });
+    },
+    logout(state) {
+      return Object.assign(state, {
+        user: null,
+        token: null,
+      });
+    },
+  },
+});
+
+export const selectAuth = (state: RootState) => state.auth;
+export const selectUser = (state: RootState) => selectAuth(state).user;
+export const selectToken = (state: RootState) => selectAuth(state).token;
+
+export const { login, logout } = authSlice.actions;
+
+export const credentialStorage = createListenerMiddleware();
```

Comment:
Yup! They make it really easy to build middleware in, and this is where any side effects of actions should be specified (you don't want to do that in reducers, since they need to be pure and not affect any external state whatsoever).